modules/client/src/internal/client/methods.ts

```diff
@@ -276,77 +275,49 @@ export class ClientMethods extends ClientCore implements IClientMethods {
-      const tokenInstance = new Contract(
-        tokenAddress,
-        erc20ContractAbi,
-        signer,
-      );
-      const currentAllowance = await tokenInstance.allowance(
-        await signer.getAddress(),
-        daoAddress,
-      );
-      yield {
-        key: DaoDepositSteps.CHECKED_ALLOWANCE,
-        allowance: currentAllowance.toBigInt(),
-      };
-      // if its lower than the needed, set it to the correct one
-      if (currentAllowance.lt(params.amount)) {
-        // If the target is an ERC20 token, ensure that the amount can be transferred
-        // Relay the yield steps to the caller as they are received
-        yield* this.setAllowance(
-          {
-            amount: params.amount,
-            spender: daoAddress,
+    switch (params.type) {
+      case TokenType.ERC20 || TokenType.NATIVE:
+        // check and update allowance if needed
+        if (
+          params.type === TokenType.ERC20 && params.tokenAddress &&
+          params.tokenAddress !== AddressZero
+        ) {
+          const { tokenAddress, daoAddressOrEns } = params;
+          // check current allowance
+          const tokenInstance = new Contract(
             tokenAddress,
-          },
-        );
-      }
-    }
-
-    // Doing the transfer
-    const daoInstance = DAO__factory.connect(daoAddress, signer);
-    const override: { value?: bigint } = {};
-
-    if (tokenAddress === AddressZero) {
-      // Ether
-      override.value = amount;
-    }
-
-    const tx = await daoInstance.deposit(
-      tokenAddress,
-      amount,
-      reference,
-      override,
-    );
-    yield { key: DaoDepositSteps.DEPOSITING, txHash: tx.hash };
-
-    const cr = await tx.wait();
-    const log = findLog(cr, daoInstance.interface, "Deposited");
-    if (!log) {
-      throw new FailedDepositError();
-    }
-
-    const daoInterface = DAO__factory.createInterface();
-    const parsedLog = daoInterface.parseLog(log);
-
-    if (!amount.toString() === parsedLog.args["amount"]) {
-      throw new AmountMismatchError(
-        amount,
-        parsedLog.args["amount"].toBigInt(),
-      );
+            ERC20_ABI,
+            signer,
+          );
+          const currentAllowance = await tokenInstance.allowance(
+            await signer.getAddress(),
+            daoAddressOrEns,
+          );
+          yield {
+            key: DaoDepositSteps.CHECKED_ALLOWANCE,
+            allowance: currentAllowance.toBigInt(),
+          };
+          // if its lower than the needed, set it to the correct one
+          if (currentAllowance.lt(params.amount)) {
+            // If the target is an ERC20 token, ensure that the amount can be transferred
+            // Relay the yield steps to the caller as they are received
+            yield* this.setAllowance(
+              {
+                amount: params.amount,
+                spender: daoAddressOrEns,
+                tokenAddress,
+              },
+            );
+          }
+        }
```

Comment:
Wouldn't you also move this to an internal helper function?

Reply:
Yes the problem is that his calls `this.setAllowance` so i'd nedd to create an internal function inside the class insted o creating it in utils, sinde it was a small piece of code i left it there

Reply:
What is the issue with the call to setAllowance? You can link one iterator to the call of another iterator as we are doing already

Reply:
I abstracted getting the allowance but I cannot abstract setting the allowance because it uses this, and if I pass it as a function it loses the context, the other option would be abstracting also the setAllowance